### PR TITLE
West Virginia Social Security Benefits Subtraction

### LINKED ISSUE #2682: West Virginia social security benefits subtractions
Thanks for filing an issue! Please remove any top-level headers that do not apply.

https://tax.wv.gov/Documents/PIT/2022/PersonalIncomeTaxFormsAndInstructions.2022.pdf#Page=25

For taxable years beginning on and after January 1, 2022, 100 percent of the amount of social security benefits received and
included in federal adjusted gross income for the taxable year shall be allowed as a decreasing modification from federal adjusted gross income when determining West Virginia taxable income subject to the tax imposed by this article, subject to the W. Va. Code §11-21-12(c). The deduction may be claimed only when the federal adjusted gross income of a married couple filing a joint return does not exceed $100,000, or $50,000 in the case of a single (including Head of Household, and Widow(er)) individual or a married individual filing a separate return).

### PR DESCRIPTION
**Pull request by qingqiuzhang**
Fixes #2682

### Summary
📝📊🧪

This pull request adds a new feature to the policy engine that allows for the calculation of the West Virginia social security benefits subtractions, which reduce the state income tax for eligible taxpayers. It introduces new parameters, variables, and tests for the feature, and updates the changelog entry accordingly. It also reuses the existing code for the household state income tax variable to incorporate the subtractions.

> _`West Virginia` code_
> _Adding deduction logic_
> _Autumn of taxes_

### Walkthrough
*  Bump minor version and add changelog entry for West Virginia social security benefits subtractions feature ([link](https://github.com/PolicyEngine/policyengine-us/pull/2954/files?diff=unified&w=0#diff-5885042762833a4aef51405a4036067e4e6aefa7a5091a2987d1d9d1fc2647e8R1-R4))
*  Add new parameter file for income threshold for West Virginia social security benefits subtractions, with metadata, values, and references ([link](https://github.com/PolicyEngine/policyengine-us/pull/2954/files?diff=unified&w=0#diff-07b7b4f0c51b1d62ebcd335dcae78547ed7e7c1f4c5080b09bb0944c83f33cb3R1-R25))
*  Add new variable files for West Virginia social security deduction and eligibility, with value types, entities, labels, units, definition periods, references, state codes, and formulas ([link](https://github.com/PolicyEngine/policyengine-us/pull/2954/files?diff=unified&w=0#diff-d1f49e91bb71742b9edbf92bde5318b64a980f5d8bb3d89536b142b23f52c90cR1-R18), [link](https://github.com/PolicyEngine/policyengine-us/pull/2954/files?diff=unified&w=0#diff-06542ec499bc5a610ef320f6aa08bf03f386b6ab5ebeeea5e844d6651b51f5aaL1-R22))
*  Add new test file for West Virginia social security benefits subtractions, with six test cases and expected outputs ([link](https://github.com/PolicyEngine/policyengine-us/pull/2954/files?diff=unified&w=0#diff-8a1f015ee7c003d053f16a6b0e7efe3c311204db78d886d6032c1888b70c2bfcR1-R59))
*  Reuse existing code for household state income tax variable, which aggregates state income tax across tax units in the household ([link](https://github.com/PolicyEngine/policyengine-us/pull/2954/files?diff=unified&w=0#diff-7866c3345c8faae089db272d4e5d135c90597066ec1e2a3678d7a195bf47cd6dL74-L73))

### CODE REVIEW

File: policyengine_us/parameters/gov/states/wv/tax/income/subtractions/social_security_benefits/rate.yaml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+description: West Virginia issues social security benefits subtraction with these rates
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/wv/tax/income/subtractions/social_security_benefits/rate.yaml:1]:
```suggestion
description: West Virginia multiplies the social security benefits by this rate, under the social security benefits subtraction.
```

File: policyengine_us/parameters/gov/states/wv/tax/income/subtractions/social_security_benefits/rate.yaml [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+metadata:
+  period: year
+  unit: /1
+  label: West Virginia social security benefits subtraction rates
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/wv/tax/income/subtractions/social_security_benefits/rate.yaml:5]:
```suggestion
  label: West Virginia social security benefits subtraction rate
```

File: policyengine_us/parameters/gov/states/wv/tax/income/subtractions/social_security_benefits/threshold.yaml
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/wv/tax/income/subtractions/social_security_benefits/threshold.yaml:1]: rename to income_threshold

File: policyengine_us/variables/gov/states/wv/tax/income/subtractions/wv_social_security_benefits_subtraction.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        "https://code.wvlegislature.gov/11-21-12/",
+    )
+    defined_for = "wv_social_security_benefits_subtraction_eligible"
+
+    def formula(tax_unit, period, parameters):
+        p = parameters(
+            period
+        ).gov.states.wv.tax.income.subtractions.social_security_benefits
+        person = tax_unit.members
+        amount = person("social_security", period) * p.rate
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/wv/tax/income/subtractions/wv_social_security_benefits_subtraction.py:27]:
```suggestion
        social_security = add(tax_unit, period ["social_security", period])
```

File: policyengine_us/variables/gov/states/wv/tax/income/subtractions/wv_social_security_benefits_subtraction.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+        return tax_unit.sum(amount)
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/wv/tax/income/subtractions/wv_social_security_benefits_subtraction.py:28]:
```suggestion
        return  social_security * p.rate
```

Comment (PavelMakarchuk) [reply policyengine_us/variables/gov/states/wv/tax/income/subtractions/wv_social_security_benefits_subtraction.py:28]: I am reading the form as all social security benefits are multiplied by 35% and not individually - shouldn't change much on the output, let me know if I am wrong

File: policyengine_us/variables/gov/states/wv/tax/income/subtractions/wv_social_security_benefits_subtraction_eligible.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
+from policyengine_us.model_api import *
+
+
+class wv_social_security_benefits_subtraction_eligible(Variable):
+    value_type = float
+    entity = TaxUnit
+    label = "West Virginia social security deduction eligible"
+    unit = USD
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/wv/tax/income/subtractions/wv_social_security_benefits_subtraction_eligible.py:8]:
```suggestion
```

File: policyengine_us/variables/gov/states/wv/tax/income/subtractions/wv_social_security_benefits_subtraction_eligible.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
+from policyengine_us.model_api import *
+
+
+class wv_social_security_benefits_subtraction_eligible(Variable):
+    value_type = float
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/wv/tax/income/subtractions/wv_social_security_benefits_subtraction_eligible.py:5]:
```suggestion
    value_type = bool
```

File: policyengine_us/variables/gov/states/wv/tax/income/subtractions/wv_social_security_benefits_subtraction_eligible.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    entity = TaxUnit
+    label = "West Virginia social security deduction eligible"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/wv/tax/income/subtractions/wv_social_security_benefits_subtraction_eligible.py:7]:
```suggestion
    label = "Eligible for the West Virginia social security deduction"
```

File: policyengine_us/variables/gov/states/wv/tax/income/subtractions/wv_social_security_benefits_subtraction.py (hunk 4 above)
Comment (qingqiuzhang) [reply policyengine_us/variables/gov/states/wv/tax/income/subtractions/wv_social_security_benefits_subtraction.py:28]: I saw that in the form the security benefits are calculated for taxpayer and spouse separately.
Screenshot 2023-09-11 at 6 51 08 PM



And they are added together when calculate total subtractions. So I do it this way.
Screenshot 2023-09-11 at 6 51 34 PM

Comment (PavelMakarchuk) [reply policyengine_us/variables/gov/states/wv/tax/income/subtractions/wv_social_security_benefits_subtraction.py:28]: Is it only for head and spouse? If so we should specify that (exclude dependents)

Comment (PavelMakarchuk) [reply policyengine_us/variables/gov/states/wv/tax/income/subtractions/wv_social_security_benefits_subtraction.py:28]: But we can still consolidate that as head and spouse ss multiplied by 35%

**Review by PavelMakarchuk — APPROVED** [review]

**Comment by MaxGhenis** [comment]
Looking at the law I think this should just be taxable ss benefits right? May also include SSI?

**Comment by qingqiuzhang** [comment]
> Looking at the law I think this should just be taxable ss benefits right? May also include SSI?
fixed it.

**Comment by qingqiuzhang** [comment]
> Looking at the law I think this should just be taxable ss benefits right? May also include SSI?

Fixed it.

File: changelog_entry.yaml [hunk 8]
```diff
@@ -0,0 +1,4 @@
+- bump: minor
+  changes:
+    fixed:
```
Comment (PavelMakarchuk) [inline changelog_entry.yaml:3]:
```suggestion
    added:
```

File: policyengine_us/parameters/gov/states/wv/tax/income/subtractions/social_security_benefits/income_threshold.yaml [hunk 9]
```diff
@@ -0,0 +1,27 @@
+description: West Virginia provides social security benefits subtraction for filers with adjusted gross income at or below this amount.
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/wv/tax/income/subtractions/social_security_benefits/income_threshold.yaml:1]:
```suggestion
description: West Virginia qualifies filers for the social security benefits subtraction with adjusted gross income at or below this amount.
```

File: policyengine_us/variables/gov/states/wv/tax/income/subtractions/wv_social_security_benefits_subtraction.py [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    def formula(tax_unit, period, parameters):
+        p = parameters(
+            period
+        ).gov.states.wv.tax.income.subtractions.social_security_benefits
+        person = tax_unit.members
+
+        is_head = person("is_tax_unit_head", period)
+        is_spouse = person("is_tax_unit_spouse", period)
+
+        head_or_spouse = is_head + is_spouse
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/wv/tax/income/subtractions/wv_social_security_benefits_subtraction.py:31]:
```suggestion
        head_or_spouse = person("is_tax_unit_head_or_spouse", period)
```

File: policyengine_us/variables/gov/states/wv/tax/income/subtractions/wv_social_security_benefits_subtraction_eligible.py [hunk 11]
```diff
@@ -0,0 +1,27 @@
+from policyengine_us.model_api import *
+
+
+class wv_social_security_benefits_subtraction_eligible(Variable):
+    value_type = bool
+    entity = TaxUnit
+    label = "Eligible for the West Virginia social security deduction"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/wv/tax/income/subtractions/wv_social_security_benefits_subtraction_eligible.py:7]:
```suggestion
    label = "Eligible for the West Virginia social security subtraction"
```

File: policyengine_us/variables/gov/states/wv/tax/income/subtractions/wv_social_security_benefits_subtraction.py [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+    defined_for = "wv_social_security_benefits_subtraction_eligible"
+
+    def formula(tax_unit, period, parameters):
+        p = parameters(
+            period
+        ).gov.states.wv.tax.income.subtractions.social_security_benefits
+        person = tax_unit.members
+
+        head_or_spouse = person("is_tax_unit_head_or_spouse", period)
+        social_security = tax_unit.sum(
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/wv/tax/income/subtractions/wv_social_security_benefits_subtraction.py:29]: please line split this comment. also for future reference, `add(person, period, ["taxable_social_security", "ssi"])` could have been more efficient for the previous logic
```suggestion
        # "social security benefits paid by the Social Security Administration as Old Age, Survivors and Disability Insurance Benefits as provided in §42 U.S.C. 401 et. seq. or as Supplemental Security Income for the Aged, Blind, and Disabled as provided in §42 U.S.C. 1381 et. seq., included in federal adjusted gross income"
        # SSI is not included in federal AGI, so limit to taxable SS.
        taxable_ss = person("taxable_social_security", period)
        total_eligible_ss = tax_unit.sum(taxable_ss * head_or_spouse)
        return total_eligible_ss * p.rate
```

File: policyengine_us/parameters/gov/states/wv/tax/income/subtractions/social_security_benefits/income_threshold.yaml
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/wv/tax/income/subtractions/social_security_benefits/income_threshold.yaml:1]: rename `income_limit.yaml`

File: policyengine_us/parameters/gov/states/wv/tax/income/subtractions/social_security_benefits/income_threshold.yaml [hunk 13]
```diff
@@ -0,0 +1,27 @@
+description: West Virginia qualifies filers for the social security benefits subtraction with adjusted gross income at or below this amount.
```
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/wv/tax/income/subtractions/social_security_benefits/income_threshold.yaml:1]:
```suggestion
description: West Virginia limits its social security benefits subtraction to filers with federal adjusted gross income at or below this amount.
```

File: policyengine_us/parameters/gov/states/wv/tax/income/subtractions/social_security_benefits/income_threshold.yaml [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+
+metadata:
+  period: year
+  unit: currency-USD
+  label: West Virginia social security benefits subtraction income threshold
```
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/wv/tax/income/subtractions/social_security_benefits/income_threshold.yaml:6]:
```suggestion
  label: West Virginia social security benefits subtraction income limit
```

File: policyengine_us/parameters/gov/states/wv/tax/income/subtractions/social_security_benefits/rate.yaml [hunk 15]
```diff
@@ -0,0 +1,18 @@
+description: West Virginia multiplies the social security benefits by this rate, under the social security benefits subtraction.
```
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/wv/tax/income/subtractions/social_security_benefits/rate.yaml:1]:
```suggestion
description: West Virginia subtracts this fraction of taxable social security benefits from adjusted gross income.
```

File: changelog_entry.yaml [hunk 16]
```diff
@@ -0,0 +1,4 @@
+- bump: minor
+  changes:
+    added:
+      - West Virginia social security benefits subtractions.
```
Comment (MaxGhenis) [inline changelog_entry.yaml:4]:
```suggestion
      - West Virginia social security benefits subtraction.
```

**Review by MaxGhenis — CHANGES_REQUESTED** [review]

File: policyengine_us/variables/gov/states/wv/tax/income/subtractions/wv_social_security_benefits_subtraction.py (hunk 12 above)
Comment (MaxGhenis) [reply policyengine_us/variables/gov/states/wv/tax/income/subtractions/wv_social_security_benefits_subtraction.py]: Please do not resolve a suggestion comment without accepting it or explaining why the suggestion is incorrect

File: policyengine_us/variables/gov/states/wv/tax/income/subtractions/wv_social_security_benefits_subtraction_eligible.py [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+    )
+    defined_for = StateCode.WV
+
+    def formula(tax_unit, period, parameters):
+        filing_status = tax_unit("filing_status", period)
+        agi = tax_unit("adjusted_gross_income", period)
+        p = parameters(
+            period
+        ).gov.states.wv.tax.income.subtractions.social_security_benefits
+        return agi <= p.income_threshold[filing_status]
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/wv/tax/income/subtractions/wv_social_security_benefits_subtraction_eligible.py:29]:
```suggestion
        return agi <= p.income_limit[filing_status]
```

File: policyengine_us/variables/gov/states/wv/tax/income/subtractions/wv_social_security_benefits_subtraction.py (hunk 12 above)
Comment (qingqiuzhang) [reply policyengine_us/variables/gov/states/wv/tax/income/subtractions/wv_social_security_benefits_subtraction.py]: Sorry, I'm not sure if you were commenting on a outdated commit. I've modified the code according to the suggestion already.

**Review by PavelMakarchuk — APPROVED** [review]

File: policyengine_us/variables/gov/states/wv/tax/income/subtractions/wv_social_security_benefits_subtraction.py [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+    def formula(tax_unit, period, parameters):
+        p = parameters(
+            period
+        ).gov.states.wv.tax.income.subtractions.social_security_benefits
+        person = tax_unit.members
+        head_or_spouse = person("is_tax_unit_head_or_spouse", period)
+        # Federal adjusted gross income includes:
+        # 1. social security benefits paid by the Social Security Administration as Old Age, Survivors and Disability Insurance Benefits
+        # 2. Supplemental Security Income for the Aged, Blind, and Disabled
+        taxable_ss = add(person, period, ["taxable_social_security", "ssi"])
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/wv/tax/income/subtractions/wv_social_security_benefits_subtraction.py:31]: The law states SSI only to the degree it's counted in AGI, and AGI does not currently include SSI.
```suggestion
        taxable_ss = person("taxable_social_security", period)
```

File: policyengine_us/tests/policy/baseline/gov/states/wv/tax/income/subtractions/wv_social_security_benefits_subtraction.yaml
Comment (MaxGhenis) [inline policyengine_us/tests/policy/baseline/gov/states/wv/tax/income/subtractions/wv_social_security_benefits_subtraction.yaml:1]: Can you reduce the number of tests? This is a fairly simple formula.

**Review by MaxGhenis — CHANGES_REQUESTED** [review]

**Review by MaxGhenis — APPROVED** [review]
good to merge after formatting fix

**Pull request merged**